### ISSUE TITLE
Kernel wakelocks (Alarmtimer and Netmgr_wl) when device is on mobile data

### ISSUE THREAD

**Issue by ameshkov** [AdguardTeam/AdguardForAndroid#1592]
User report:
https://forum.xda-developers.com/showpost.php?p=74329111&postcount=2824

We need to do an additional investigation of this issue.

Basically, we need to collect and compare battery stats with and without AdGuard.

Here is how to collect necessary stats:
1. Reset stats: `adb shell dumpsys batterystats --reset`
2. Use the device for a fixed period of time (at least an hour)
3. Dump the battery stats: ` adb shell dumpsys batterystats > batterystats.txt`

We will need two `batterystats.txt` files. One recorded with AdGuard and the second one recorded without without AG.

**Comment by MR-Diamond** [comment]
Seems that I found the same issue; would be enough provide battery's stats from Better Battery Stats? This app gives detailed stats.

**Comment by ameshkov** [comment]
@MR-Diamond 
Could you please export stats from BBS and attach? I'll check if it works for us

**Comment by MR-Diamond** [comment]
Ok, here 2 logs: one with AdGuard enabled and the second one with AdGuard disabled.

I used the device for a fixed period of time (less than an hour because currently I am far from home and I forget to take the charger) and by doing the same things, by using the same apps in the same way. I am on mobile data.

The  first noticeable things: 
>(**AdGuard disabled**)
>10287 (Mobile) (com.adguard.android.AdGuard): 139.0 KBytes 4,3%”
>Awake (Screen Off) (): 7 m 7 s  16,8%
>Level lost [%]: Bat.: -3%(51% to 48%) [4,3%/h]

>(**AdGuard enabled**)
>10287 (Mobile) (com.adguard.android.AdGuard): 21.0 MBytes 98,5%
>Awake (Screen Off) (): 10 m 20 s  24,9%
>Level lost [%]: Bat.: -4%(48% to 44%) [5,8%/h]

About 10287 (Mobile) (com.adguard.android.AdGuard): 21.0 MBytes 98,5%:
I think is normal, since it filters ads from apps: in the previous logs, in the Network section, in facts we can see mobile data used by each app, but by judging from this log - AdGuard enabled - not all apps seems that has been filtered by AdGuard; however I don't understand why, when was disabled, it used a little bit the connection.

[AdGuard_enabled_BetterBatteryStats-2017-11-17_143103189.txt](https://github.com/AdguardTeam/AdguardForAndroid/files/1482444/AdGuard_enabled_BetterBatteryStats-2017-11-17_143103189.txt)
[AdGuard_disabled_BetterBatteryStats-2017-11-17_134706161.txt](https://github.com/AdguardTeam/AdguardForAndroid/files/1482445/AdGuard_disabled_BetterBatteryStats-2017-11-17_134706161.txt)

**Comment by ameshkov** [comment]
> when was disabled, it used a little bit the connection.

It might've checked the app or filters updates.

**Comment by ameshkov** [comment]
Yeah, this type of stats works for us. But we need them for a greater period (the whole day for instance). Otherwise, it's hard to make any conclusions.

**Comment by MR-Diamond** [comment]
>It might've checked the app or filters updates.

In AdGuard (since days) I disabled check for updates and send statistics, since I already noticed that It used mobile data when inactive.

>Yeah, this type of stats works for us. But we need them for a greater period (the whole day for instance). Otherwise, it's hard to make any conclusions.

You're right: I the next days I will see to provide you logs for a longer time.

**Comment by ameshkov** [comment]
> In AdGuard (since days) I disabled check for updates and send statistics, since I already noticed that It used mobile data when inactive.

If you have AG "record everything" log enabled, it should be visible there what's being done.

**Comment by MR-Diamond** [comment]
Sorry, in the meantime I have a question:

Using [Nmap](https://nmap.org/) (a tool to discover open ports on hosts) I found that my device (I have a public IP) has an open port.

I used: nmap -T4 -A -v -Pn MYIPADDRESS

I can say for sure that is opened by AdGuard; I ping my IP on such port and the ping works: if I stop and put AdGuard in Greenify I can no longer ping my IP, I have tried many times and the result il always the same. Why?

**Comment by ameshkov** [comment]
Open? It's the port AdGuard's proxy is listening on

**Comment by worstperson** [comment]
I think he's saying the port is accessible over the internet when his device is given a public IP(no nat, no router), were you'd expect it to be restricted to the device or local net.

**Comment by MR-Diamond** [comment]
@worstperson , yes: is what I meant/said.

**Comment by ameshkov** [comment]
It does listen on 127.0.0.1 so in theory, it is accessible from anywhere when you have a public IP address.

In case of the VPN or Proxy+Auto filtering mode, AG should not process that connection as it won't be able to detect what app caused that connection.

In case of "proxy with manual setup" AG will process any incoming connection, so you can even set your other devices to use AG as an upstream proxy.

**Comment by MR-Diamond** [comment]
@ameshkov thank you for the explanation.
Anyway, since my phone ain't rooted, I use AdGuard in VPN mode; there is a way to change the port where AdGuard is listen, or better close/block such port?

**Comment by ameshkov** [comment]
> Anyway, since my phone ain't rooted, I use AdGuard in VPN mode; there is a way to change the port where AdGuard is listen, or better close/block such port?

In VPN mode it just looks for a first available port. VPN should prevent all the outside connections anyway.

**Comment by worstperson** [comment]
That's what I thought was weird. I'd think adguard would bind to lo as you said, but I'm pretty sure it has to be bound to that separate interface for it to open a port as reported.

**Comment by ameshkov** [comment]
> That's what I thought was weird. I'd think adguard would bind to lo as you said, but I'm pretty sure it has to be bound to that separate interface for it to open a port as reported.

No idea:(

Basically, the code for that is this:
```
InetSocketAddress listenAddress = new InetSocketAddress("127.0.0.1", AvailablePortFinder.getNextAvailable());
...
...
serverSocketChannel.socket().bind(listenAddress);
```

**Comment by MR-Diamond** [comment]
I back on topic.

I also tried Blokada but I still found the same battery drain.
This time I saved a BBS log after 12 hours and 25 minutes.

Someone, by looking at this log have some idea about the culprit?
The most evident wakelock is `alarmtimer   (*api*): 1 h 3 m 53 s  Cnt:(c/wc/ec)1921/0/0 8,6%` which is almost equal to `Awake (Screen Off) (): 1 h 14 m 24 s  10,0%`, but I can't figure out from which app or service is triggered. In the sync section of Android I turned off almost everything.

[BetterBatteryStats-2017-11-20_104916471.txt](https://github.com/AdguardTeam/AdguardForAndroid/files/1487397/BetterBatteryStats-2017-11-20_104916471.txt)

**Comment by ameshkov** [comment]
@MR-Diamond do you have the same high result without ad blocking?

**Comment by MR-Diamond** [comment]
Nope: if I don't use any ad blocker I don't face any excessive battery drain.
I attach another BBS log after 11 hours without adblock (AdGuard) active and this time I just lose -3% instead of -10%.

usually for me, to notice this battery drain, also 30 mins of idle are enough, thanks to the BBS log.
What I noticed is that the most of the time, the top offender is Google Play Services.

After 30 mins of idle in the Alarms section, I see at least 10 wakeups from com.google.android.gms; I also tried with AdGuard disabled (and greenified), and in such case (after 30 mins), I don't see (or I see less) wakeups from com.google.android.gms.

What can I try to do?  Exclude Google Play Services from AdGuard? Would be enough? Or is AdGuard which calls Google Play Services, as applications usually do?

**Comment by ameshkov** [comment]
I wouldn't advise disabling AG for `com.google.android.gms` as it is used to load ads.

> I attach another BBS log after 11 hours without adblock (AdGuard) active and this time I just lose -3% instead of -10%.

Meh, the problem with it is that I see difference in the wakelocks, but I can't see what and when did it happen. 

The only possible way is to collect two logs (from AdGuard with all the requests recorded) and the one dumped with `adb` (as I described above). Then I can "map" wakelocks/alarms and requests blocked by AG and maybe come up with something.

Here is how to collect AdGuard's log properly (logging level should be switched prior to the test):
https://kb.adguard.com/en/android/solving-problems/log

**Comment by MR-Diamond** [comment]
Ok, I collected the two logs which you asked me.
This time, in about 11 hours (with AdGuard enabled) I lost "only" -6% (from 55% to 49%) and I think is thanks to the fact that I restricted background data to a lot of applications; usually in 11 hours (without any ad blocker installed) I loose about -2% or -3%.

About the logs: maybe there is indicated my IP address? Is public, and I'd want to avoid to reveal it publicly. Can I send you these logs (also the battery dump log) using E-mail? Let me know about the address where to sent these files. Thanks.

**Comment by MR-Diamond** [comment]
Ok, I found the email address here: https://kb.adguard.com/en/technical-support
I sent full logs in a zip file.

**Comment by ameshkov** [comment]
@MR-Diamond huh, you'd better send it directly to `devteam@adguard.com` to avoid the support lines.

**Comment by MR-Diamond** [comment]
Done.

**Comment by ameshkov** [comment]
@MR-Diamond got it, thank you!

**Comment by MR-Diamond** [comment]
I quicly looks the logs and I see that the port usde by proxy is 8080; the port which I find open is another, which is not reported in the state.txt file.

**Comment by ameshkov** [comment]
@MR-Diamond look for these lines:
```
Creating server listening to /127.0.0.1:1104
Starting DNS proxy server on /127.0.0.1:1105
dnscrypt-proxy has been started on /127.0.0.1:1106
````

**Comment by ameshkov** [comment]
Hm, I just realized that it is dnscrypt-proxy that might be exposing an open port

**Comment by MR-Diamond** [comment]
Nope, is another port, but again: since I want to keep my phone/IP private, I think that is better that I indicate such port privately. I will send another email.

**Comment by MR-Diamond** [comment]
Such port is open also if I disable DNS crypt.

**Comment by ameshkov** [comment]
Try using telnet to connect to this port and check if there's really anything listening there.

**Comment by MR-Diamond** [comment]
I'm not really practice with telnet. I am able to connect to my IP and by indicating such port; I have to run some other commands? Currently I am on Windows 7.

**Comment by ameshkov** [comment]
Nope, just checking if there is something there.

**Comment by MR-Diamond** [comment]
Ok: when I connect with telnet, the window is empty.

**Comment by MR-Diamond** [comment]
@ameshkov 

A quick question, after reading the batterystats  log:
I see a section called `Package changes:` here I can see some apps which I uninstalled also weeks ago (and I also wiped the phone's cache various times): why are these still listed? Not all of  these has been uninstalled (like com.google.android.apps.docs.editors.docs and  com.google.android.calendar) but the others yes: has been uninstalled but I cannot also see them using the app Package Name Viewer (well, uninstalled apps are no longer reported on my phone, but some of them are reported in the batterystats log).

Why?
